File: ui/core/wasm/bulk_sim/estimate.ts
import { BulkSimRequest, BulkSimStage } from '../../proto/api';
import { BULK_SIM_MIN_COMBINATIONS } from './constants';
import { bulkSimStageConfigs, getBulkSimStageMaxSurvivors, getBulkSimStageMinIterations, shouldRunBulkSimStage } from './stage';

export const shouldUseLegacyBulkSim = (request: BulkSimRequest, candidateCount: number): boolean => {
	const settings = request.bulkSettings;
	if (settings?.useLegacyBulkSim) {
		return true;
	}
	if (candidateCount < BULK_SIM_MIN_COMBINATIONS) {
		return true;
	}

	const highStageIterations = request.highStageIterations;
	let remainingCandidates = candidateCount;
	let estimatedOptimisationIterationsUpperBound = 0;

	for (const config of bulkSimStageConfigs) {
		if (config.stage === BulkSimStage.BulkSimStageHigh) {
			break;
		}
		if (!shouldRunBulkSimStage(config, remainingCandidates)) {
			continue;
		}

		estimatedOptimisationIterationsUpperBound += getBulkSimStageMinIterations(request, config) * (remainingCandidates + 1);
		remainingCandidates = Math.min(remainingCandidates, getBulkSimStageMaxSurvivors(config, remainingCandidates) ?? remainingCandidates);
	}

	estimatedOptimisationIterationsUpperBound +=
		getBulkSimStageMinIterations(request, bulkSimStageConfigs[bulkSimStageConfigs.length - 1]!) * (remainingCandidates + 1);
	return estimatedOptimisationIterationsUpperBound >= highStageIterations * candidateCount;
};

File: sim/core/bulk/estimate.go
package bulk

import (
	"github.com/wowsims/tbc/sim/core/proto"
)

func shouldUseLegacyBulkSim(settings *proto.BulkSettings, highStageIterations int32, candidateCount int) bool {
	if settings != nil && settings.GetUseLegacyBulkSim() {
		return true
	}
	if candidateCount < bulkSimMinCombinations {
		return true
	}

	fullRunIterations := int64(highStageIterations) * int64(candidateCount)
	estimatedMultistageIterationsUpperBound := getBulkSimOptimisationIterationsUpperBound(highStageIterations, candidateCount)
	return estimatedMultistageIterationsUpperBound >= fullRunIterations
}

func getBulkSimOptimisationIterationsUpperBound(highStageIterations int32, candidateCount int) int64 {
	remainingCandidates := candidateCount
	var iterations int64

	for _, stageConfig := range bulkSimStageConfigs {
		if stageConfig.Stage == proto.BulkSimStage_BulkSimStageHigh {
			break
		}
		if !shouldRunBulkSimStage(stageConfig, remainingCandidates) {
			continue
		}

		stageIterations := getBulkSimStageMinIterationsFromFloor(highStageIterations, stageConfig)
		iterations += int64(stageIterations) * int64(remainingCandidates+1)
		remainingCandidates = min(remainingCandidates, getBulkSimStageMaxSurvivors(stageConfig, remainingCandidates))
	}

	return iterations + int64(highStageIterations)*int64(remainingCandidates+1)
}

func estimateBulkSimIterations(settings *proto.BulkSettings, highStageIterations int32, candidateCount int) (int64, bool) {
	if shouldUseLegacyBulkSim(settings, highStageIterations, candidateCount) {
		return int64(highStageIterations) * int64(candidateCount), true
	}

	return getBulkSimOptimisationIterationsUpperBound(highStageIterations, candidateCount), false
}
